src/MultiThreadedCaches.jl

```diff
@@ -105,73 +107,124 @@ function _thread_cache(mtcache::MultiThreadedCache)
     return cache
 end
 @noinline _thread_cache_length_assert() = @assert false "** Must call `init_cache!(cache)` in your Module's __init__()! - length(cache.thread_caches) < Threads.nthreads() "
+function _thread_lock(cache::MultiThreadedCache, tid)
+    length(cache.thread_locks) >= Threads.nthreads() || _thread_cache_length_assert()
+    if @inbounds isassigned(cache.thread_locks, tid)
+        @inbounds lock = cache.thread_locks[tid]
+    else
+        lock = eltype(cache.thread_locks)()
+        @inbounds cache.thread_locks[tid] = lock
+    end
+    return lock
+end
+@noinline _thread_cache_length_assert() = @assert false "** Must call `init_cache!(cache)` in your Module's __init__()! - length(cache.thread_caches) < Threads.nthreads() "
 
 
 const CACHE_MISS = :__MultiThreadedCaches_key_not_found__
 
 function Base.get!(func::Base.Callable, cache::MultiThreadedCache{K,V}, key) where {K,V}
```

Comment:
Ugh the indentation change here down is annoying. I'd recommend hiding whitespace in the github UI:

<img width="281" alt="Screen Shot 2022-02-15 at 6 03 11 PM" src="https://user-images.githubusercontent.com/1582097/154164532-35624ac9-f9aa-497d-a93d-fb15316ce077.png">